distributed/tests/test_scheduler.py

```diff
@@ -359,33 +378,18 @@ async def test_graph_execution_width(c, s, *workers):
     The number of parallel work streams match the number of threads.
     """
 
-    class Refcount:
-        "Track how many instances of this class exist; logs the count at creation and deletion"
-
-        count = 0
-        lock = dask.utils.SerializableLock()
-        log = []
-
-        def __init__(self):
-            with self.lock:
-                type(self).count += 1
-                self.log.append(self.count)
-
-        def __del__(self):
-            with self.lock:
-                self.log.append(self.count)
-                type(self).count -= 1
```

Comment:
This test is interesting. This PR is not changing anything in terms of scheduling, ordering, etc. but this is still quite reliably failing. It seems as if Refcount is relying on explicit garbage collection. This is something I want to look into a little more since we're seeing a lot of GC warnings recently. However, for the sake of this PR I rewrote it to count keys in data instead of relying on GC. Eventually, I think both tests would make sense

Reply:
This is really a weird case and somehow connected to how this object is defined in a local context.
I looked pretty closely but I cannot find any cyclic references. In fact, I see actually _fewer_ objects actually tracked by GC than this counter is let to believe. I know that CPython guarantees that `__del__` is indeed called and only called once but I believe there are some caveats about when this is the case.